src/Internal/Logger.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Class Logger
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal;
+
+use WC_Logger;
+use WC_Payments;
+use WCPay\Core\Mode;
+use WCPay\Internal\Proxy\LegacyProxy;
+
+defined( 'ABSPATH' ) || exit; // block direct access.
+
+/**
+ * A wrapper class for interacting with WC_Logger.
+ */
+class Logger {
+
+	const LOG_FILENAME = 'woocommerce-payments';
+
+	/**
+	 * The holding property for our WC_Logger instance.
+	 *
+	 * @var WC_Logger $logger
+	 */
+	private $wc_logger;
+
+	/**
+	 * Mode
+	 *
+	 * @var Mode
+	 */
+	private $mode;
+
+	/**
+	 * Logger constructor.
+	 *
+	 * @param WC_Logger $wc_logger    Legacy proxy.
+	 * @param Mode      $mode            Mode.
+	 */
+	public function __construct( WC_Logger $wc_logger, Mode $mode ) {
+		$this->wc_logger = $wc_logger;
+		$this->mode      = $mode;
+	}
+
+	/**
+	 * Add a log entry.
+	 *
+	 * Note that this depends on WC_Payments gateway property to be initialized as
+	 * we need this to access the plugins debug setting to figure out if the setting
+	 * is turned on.
+	 *
+	 * @param string $message Log message.
+	 * @param string $level One of the following:
+	 *     'emergency': System is unusable.
+	 *     'alert': Action must be taken immediately.
+	 *     'critical': Critical conditions.
+	 *     'error': Error conditions.
+	 *     'warning': Warning conditions.
+	 *     'notice': Normal but significant condition.
+	 *     'info': Informational messages.
+	 *     'debug': Debug-level messages.
+	 * @throws \Exception Throws Exception if log cannot be written.
+	 */
+	public function log( $message, $level = 'info' ) {
+		if ( ! $this->can_log() ) {
+			return;
+		}
+
+		$this->wc_logger->log( $level, $message, [ 'source' => self::LOG_FILENAME ] );
+	}
+
+
+	/**
+	 * Checks if the gateway setting logging toggle is enabled.
+	 *
+	 * @return bool Depending on the enable_logging setting.
+	 * @throws \Exception Throws Exception for any issue in accessing mode.
+	 */
+	public function can_log() {
+
+		if ( $this->mode->is_dev() ) {
+			return true;
+		}
```

Comment:
Now that we're trying to do a better job with exceptions (you've already seen the Psalm errors), I see that the `Exception` thrown within the `Mode` class is too generic. If we kept on throwing it here, this would mean that every method, which uses the logger, will either need to handle the generic `Exception` class, or keep on throwing it, which is not ideal.

For the sake of the logger, I think it's safe to handle the exception here, and simply return false. It's a very edge case anyway.

```suggestion
		try {
			if ( $this->mode->is_dev() ) {
				return true;
			}
		} catch ( Exception $e ) {
			return false;
		}
```

Also, let's avoid using `\`s and replace them with `use` statements instead.

Reply:
Done